donkeycar/pipeline/sequence.py

```diff
@@ -1,154 +1,90 @@
-
-from typing import (Any, Callable, Generic, Iterable, Iterator, List, Sized, Tuple, TypeVar)
+from typing import Any, Callable, Generic, Iterable, Iterator, List, Sized, \
+    Tuple, TypeVar, Union
 
 from donkeycar.pipeline.types import TubRecord
 
 # Note: Be careful when re-using `TypeVar`s.
 # If you are not-consistent mypy gets easily confused.
 
-R = TypeVar('R', covariant=True)
 X = TypeVar('X', covariant=True)
 Y = TypeVar('Y', covariant=True)
 XOut = TypeVar('XOut', covariant=True)
 YOut = TypeVar('YOut', covariant=True)
 
 
-class SizedIterator(Generic[X], Iterator[X], Sized):
+# This is a protocol type without explicitly using a `Protocol`
+# Using `Protocol` requires Python 3.7
+class SizedIterable(Generic[X], Iterable[X], Sized):
     def __init__(self) -> None:
-        # This is a protocol type without explicitly using a `Protocol`
-        # Using `Protocol` requires Python 3.7
         pass
 
+    def __len__(self) -> int:
+        pass
 
-class TubSeqIterator(SizedIterator[TubRecord]):
-    def __init__(self, records: List[TubRecord]) -> None:
-        self.records = records or list()
-        self.current_index = 0
-
-    def __len__(self):
-        return len(self.records)
-
-    def __iter__(self) -> SizedIterator[TubRecord]:
-        return TubSeqIterator(self.records)
-
-    def __next__(self):
-        if self.current_index >= len(self.records):
-            raise StopIteration('No more records')
-
-        record = self.records[self.current_index]
-        self.current_index += 1
-        return record
-
-    next = __next__
+    def __iter__(self) -> Iterator[X]:
+        pass
 
 
-class TfmIterator(Generic[R, XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+class _BaseTfmIterator(Generic[X, Y, XOut, YOut], Iterator[Tuple[XOut, YOut]]):
```

Comment:
The iterator is specific to its iterable (renamed `TfmShallowList`). It has a little logic as possible, it just serves to implement the iterator protocol.